ldap/servers/slapd/ldapi.c

```diff
@@ -0,0 +1,409 @@
+/** BEGIN COPYRIGHT BLOCK
+ * Copyright (C) 2020 Red Hat, Inc.
+ * All rights reserved.
+ *
+ * License: GPL (version 3 or any later version).
+ * See LICENSE for details.
+ * END COPYRIGHT BLOCK **/
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <unistd.h>
+#include "slap.h"
+
+#if defined(ENABLE_LDAPI)
+
+#include "getsocketpeer.h"
+
+struct ldapi_mapping {
+    char *dn;
+    uid_t uid;
+    gid_t gid;
+    struct ldapi_mapping *next;
+};
+
+static struct ldapi_mapping *ldapi_mappings = NULL;
+static pthread_mutex_t dn_mapping_lock;
```

Comment:
Consider using a rwlock here instead. If we use a mutex, with heavy threads we'll serialise on the lock, but with a rwlock we wont block and allows parallel ops to continue. Also Ensure if you use a rwlock to set it to "writer preferring" so that updates to this do not stall forever which is a problem on the default linux "reader preferring".

Reply:
Done.